src/librustc_parse/parser/ty.rs

```diff
@@ -276,8 +276,33 @@ impl<'a> Parser<'a> {
     }
 
     fn parse_borrowed_pointee(&mut self) -> PResult<'a, TyKind> {
-        let opt_lifetime = if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
+        let mut opt_lifetime =
+            if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
         let mutbl = self.parse_mutability();
+        if mutbl == Mutability::Mut && opt_lifetime.is_none() && self.token.is_lifetime() {
+            // A lifetime is invalid here: it would be part of a bare trait bound, which requires
+            // it to be followed by a plus, but we disallow plus in the pointee type.
+            // So we can handle this case as an error here, and suggest `'a mut`.
+            // If there *is* a plus next though, handling the error later provides better suggestions
+            // (like adding parentheses)
+            if !self.look_ahead(1, |t| t.is_like_plus()) {
+                let span = self.prev_token.span.to(self.token.span);
+                let lifetime = self.expect_lifetime();
+
+                let mut err = self.struct_span_err(span, "lifetime must precede `mut`");
+                if self.span_to_snippet(span).is_ok() {
+                    err.span_suggestion(
+                        span,
+                        "place the lifetime before `mut`",
+                        format!("{} mut", lifetime),
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+                err.emit();
+
+                opt_lifetime = Some(lifetime);
```

Comment:
Can you skip the lifetime token here? This should prevent the IMO unhelpful later errors here.